File: trunk/Geometry/face_height.cpp
#ifndef FACEHEIGHT_H_
#define FACEHEIGHT_H_

#include <map>
#include <new>
using namespace std;

#include "geoquant.h"
#include "triposition.h"
#include "miscmath.h"
#include "geoquants.h"

FaceHeightIndex* FaceHeight::Index = NULL;

FaceHeight::FaceHeight( Face& f, Tetra& t ){
  StdTetra st = labelTetra( f, t );

  hij_l = EdgeHeight::At(st.e12, st.f123);
  hij_k = EdgeHeight::At(st.e12, st.f124);
  beta_ij_kl = DihedralAngle::At(st.e12, t);

  hij_l->addDependent(this);
  hij_k->addDependent(this);
  beta_ij_kl->addDependent(this);
}

void FaceHeight::recalculate(){
    double Hij_l = hij_l->getValue();
    double Hij_k = hij_k->getValue();
    double angle = beta_ij_kl->getValue();

    value = Hij_l - Hij_k * cos(angle)/sin(angle);
}

void FaceHeight::remove() {
     deleteDependents();
     hij_l->removeDependent(this);
     hij_k->removeDependent(this);
     beta_ij_kl->removeDependent(this);
     Index->erase(pos);  
     delete this;
}

FaceHeight::~FaceHeight(){ }

FaceHeight* FaceHeight::At( Face& f, Tetra& t ){
  TriPosition T( 2, f.getSerialNumber(), t.getSerialNumber() );
  if( Index == NULL ) Index = new FaceHeightIndex();
  FaceHeightIndex::iterator iter = Index->find( T );

  if( iter == Index->end() ){
    FaceHeight* val = new FaceHeight( f, t );
    val->pos = T;
    Index->insert( make_pair( T, val ) );
    return val;
  } else {
    return iter->second;
  }
}

void FaceHeight::CleanUp(){
  if( Index == NULL ) return;
  FaceHeightIndex::iterator iter;
  FaceHeightIndex copy = *Index;
  for(iter = copy.begin(); iter != copy.end(); iter++) {
    iter->second->remove();
  }
    
  delete Index;
  Index = NULL;
}

#endif /* FACEHEIGHT_H_ */

File: Java/src/Geoquant/FaceHeight.java
package Geoquant;

import java.util.HashMap;

import triangulation.Face;
import triangulation.StdTetra;
import triangulation.Tetra;


public class FaceHeight extends Geoquant {
  // Index map
  private static HashMap<TriPosition, FaceHeight> Index = new HashMap<TriPosition, FaceHeight>();
  
  // Needed geoquants
  private EdgeHeight hij_l;
  private EdgeHeight hij_k;
  private DihedralAngle beta_ij_kl;
  
  public FaceHeight(Face f, Tetra t) {
    super(f, t);
    
    StdTetra st = new StdTetra(t, f);
    hij_k = EdgeHeight.At(st.e12, st.f123);
    hij_k.addObserver(this);
    hij_l = EdgeHeight.At(st.e12, st.f124);
    hij_l.addObserver(this);
    beta_ij_kl = DihedralAngle.At(st.e12, t);
    beta_ij_kl.addObserver(this);
  }
  
  protected void recalculate() {
    double Hij_l = hij_l.getValue();
    double Hij_k = hij_k.getValue();
    double angle = beta_ij_kl.getValue();

    value = (Hij_l - Hij_k * Math.cos(angle))/Math.sin(angle);
  }
 
  public void remove() {
    deleteDependents();
    hij_l.deleteObserver(this);
    hij_k.deleteObserver(this);
    beta_ij_kl.deleteObserver(this);
    Index.remove(pos);
  }
  
  public static FaceHeight At(Face f, Tetra t) {
    TriPosition T = new TriPosition(f.getSerialNumber(), t.getSerialNumber());
    FaceHeight q = Index.get(T);
    if(q == null) {
      q = new FaceHeight(f, t);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static double valueAt(Face f, Tetra t) {
    return At(f, t).getValue();
  }

}
